diff -urN civicrm-4.2.19/CRM/Core/OptionGroup.php civicrm-4.2.19-wmf/CRM/Core/OptionGroup.php

Status: needs upstreaming in some form.
Background: This introduces caching on CRM_Core_OptionGroup::getLabel
It's likely the correct fix is to deprecate that function.
JIRA: CRM-17120

--- civicrm-4.2.19/CRM/Core/OptionGroup.php	2014-09-09 15:40:20.000000000 -0700
+++ civicrm-4.2.19-wmf/CRM/Core/OptionGroup.php	2015-03-11 13:31:10.298248476 -0700
@@ -34,6 +34,7 @@
  */
 class CRM_Core_OptionGroup {
   static $_values = array();
+  static $_value_cache = array();
 
   /*
      * $_domainIDGroups array maintains the list of option groups for whom
@@ -279,6 +280,11 @@
       return NULL;
     }
 
+    $cache_key = "$groupName-$label-$valueField";
+    if ( array_key_exists( $cache_key, self::$_value_cache ) ) {
+      return self::$_value_cache[ $cache_key ];
+    }
+
     $query = "
 SELECT  v.label as label ,v.{$valueField} as value
 FROM   civicrm_option_value v,
@@ -295,8 +301,10 @@
     );
     $dao = CRM_Core_DAO::executeQuery($query, $p);
     if ($dao->fetch()) {
+      $value = $dao->value;
       $dao->free();
-      return $dao->value;
+      self::$_value_cache[ $cache_key ] = $value;
+      return $value;
     }
     $dao->free();
     return NULL;
