diff -urN civicrm-4.2.19/CRM/ACL/BAO/ACL.php civicrm-4.2.19-wmf/CRM/ACL/BAO/ACL.php
--- civicrm-4.2.19/CRM/ACL/BAO/ACL.php	2014-09-09 15:40:20.000000000 -0700
+++ civicrm-4.2.19-wmf/CRM/ACL/BAO/ACL.php	2015-03-11 13:31:10.174248480 -0700
@@ -502,7 +503,7 @@
     $rule->query($query);
 
     while ($rule->fetch()) {
-      $results[$rule->id] = &$rule->toArray();
+      $results[$rule->id] = $rule->toArray();
     }
 
     return $results;
@@ -542,7 +543,7 @@
       $rule->query($query);
 
       while ($rule->fetch()) {
-        $results[$rule->id] = &$rule->toArray();
+        $results[$rule->id] = $rule->toArray();
       }
     }
 
@@ -601,7 +602,7 @@
     $rule->query($query);
 
     while ($rule->fetch()) {
-      $results[$rule->id] = &$rule->toArray();
+      $results[$rule->id] = $rule->toArray();
     }
 
     // also get all acls for "Any Role" case
@@ -621,7 +622,7 @@
 
     $rule->query($query);
     while ($rule->fetch()) {
-      $results[$rule->id] = &$rule->toArray();
+      $results[$rule->id] = $rule->toArray();
     }
 
     return $results;
diff -urN civicrm-4.2.19/CRM/Core/Block.php civicrm-4.2.19-wmf/CRM/Core/Block.php
--- civicrm-4.2.19/CRM/Core/Block.php	2014-09-09 15:40:20.000000000 -0700
+++ civicrm-4.2.19-wmf/CRM/Core/Block.php	2015-03-11 13:31:10.282248477 -0700
@@ -281,6 +281,7 @@
    * @return void
    * @access private
    */
+  static
   private function setTemplateValues($id) {
     switch ($id) {
       case self::CREATE_NEW:
@@ -333,6 +334,7 @@
    * @return void
    * @access private
    */
+  static
   private function setTemplateShortcutValues() {
     $config = CRM_Core_Config::singleton();
 
@@ -433,6 +435,7 @@
    * @return void
    * @access private
    */
+  static
   private function setTemplateDashboardValues() {
     static $dashboardLinks = array();
     if (CRM_Core_Permission::check('access Contact Dashboard')) {
diff -urN civicrm-4.2.19/CRM/Core/I18n/Schema.php civicrm-4.2.19-wmf/CRM/Core/I18n/Schema.php
--- civicrm-4.2.19/CRM/Core/I18n/Schema.php	2014-09-09 15:40:20.000000000 -0700
+++ civicrm-4.2.19-wmf/CRM/Core/I18n/Schema.php	2015-03-11 13:31:10.294248476 -0700
@@ -478,7 +478,7 @@
     return "CREATE OR REPLACE VIEW {$table}_{$locale} AS SELECT " . implode(', ', $cols) . " FROM {$table}";
   }
 
-  function triggerInfo(&$info, $tableName = NULL) {
+  static function triggerInfo(&$info, $tableName = NULL) {
     // get the current supported locales
     $domain = new CRM_Core_DAO_Domain();
     $domain->find(TRUE);
diff -urN civicrm-4.2.19/CRM/Utils/System.php civicrm-4.2.19-wmf/CRM/Utils/System.php
--- civicrm-4.2.19/CRM/Utils/System.php	2014-09-09 15:40:20.000000000 -0700
+++ civicrm-4.2.19-wmf/CRM/Utils/System.php	2015-03-11 13:31:10.442248473 -0700
@@ -200,7 +200,7 @@
     return $config->userSystem->url($path, $query, $absolute, $fragment, $htmlize, $frontend);
   }
 
-  function href($text, $path = NULL, $query = NULL, $absolute = TRUE,
+  static function href($text, $path = NULL, $query = NULL, $absolute = TRUE,
     $fragment = NULL, $htmlize = TRUE, $frontend = FALSE
   ) {
     $url = self::url($path, $query, $absolute, $fragment, $htmlize, $frontend);
diff -urN civicrm-4.2.19/packages/HTML/QuickForm/advcheckbox.php civicrm-4.2.19-wmf/packages/HTML/QuickForm/advcheckbox.php
--- civicrm-4.2.19/packages/HTML/QuickForm/advcheckbox.php	2014-09-09 15:40:26.000000000 -0700
+++ civicrm-4.2.19-wmf/packages/HTML/QuickForm/advcheckbox.php	2014-12-17 10:23:51.000000000 -0800
@@ -271,7 +271,7 @@
     * This element has a value even if it is not checked, thus we override
     * checkbox's behaviour here
     */
-    function exportValue(&$submitValues, $assoc)
+    function exportValue(&$submitValues, $assoc = false)
     {
         $value = $this->_findValue($submitValues);
         if (null === $value) {
diff -urN civicrm-4.2.19/packages/HTML/QuickForm/RuleRegistry.php civicrm-4.2.19-wmf/packages/HTML/QuickForm/RuleRegistry.php
--- civicrm-4.2.19/packages/HTML/QuickForm/RuleRegistry.php	2014-09-09 15:40:26.000000000 -0700
+++ civicrm-4.2.19-wmf/packages/HTML/QuickForm/RuleRegistry.php	2014-12-17 10:23:51.000000000 -0800
@@ -54,7 +54,7 @@
      * @static
      * @return    HTML_QuickForm_RuleRegistry
      */
-    function &singleton()
+    static function &singleton()
     {
         static $obj;
         if (!isset($obj)) {
diff -urN civicrm-4.2.19/packages/HTML/QuickForm.php civicrm-4.2.19-wmf/packages/HTML/QuickForm.php
--- civicrm-4.2.19/packages/HTML/QuickForm.php	2014-09-09 15:40:26.000000000 -0700
+++ civicrm-4.2.19-wmf/packages/HTML/QuickForm.php	2015-03-11 13:31:10.514248471 -0700
@@ -1320,6 +1320,7 @@
     * @param    array   $b  array which will be merged into first one
     * @return   array   merged array
     */
+    static
     function arrayMerge($a, $b)
     {
         foreach ($b as $k => $v) {
