diff -urN civicrm-4.2.19/CRM/Contribute/Form/Task.php civicrm-4.2.19-wmf/CRM/Contribute/Form/Task.php
--- civicrm-4.2.19/CRM/Contribute/Form/Task.php	2014-09-09 15:40:20.000000000 -0700
+++ civicrm-4.2.19-wmf/CRM/Contribute/Form/Task.php	2015-03-11 13:31:10.262248477 -0700
@@ -166,7 +166,7 @@
    * @return void
    * @access public
    */
-  function addDefaultButtons($title, $nextType = 'next', $backType = 'back') {
+  function addDefaultButtons($title, $nextType = 'next', $backType = 'back', $submitOnce = FALSE) {
     $this->addButtons(array(
         array(
           'type' => $nextType,
