diff -urN civicrm-4.2.19/CRM/Core/BAO/Mapping.php civicrm-4.2.19-wmf/CRM/Core/BAO/Mapping.php
--- civicrm-4.2.19/CRM/Core/BAO/Mapping.php	2014-09-09 15:40:20.000000000 -0700
+++ civicrm-4.2.19-wmf/CRM/Core/BAO/Mapping.php	2015-03-11 13:31:10.278248477 -0700
@@ -272,13 +272,12 @@
   static
   function buildMappingForm(&$form, $mappingType = 'Export', $mappingId = NULL, $columnNo, $blockCount = 3, $exportMode = NULL) {
     if ($mappingType == 'Export') {
-      $name = "Map";
       $columnCount = array('1' => $columnNo);
     }
     elseif ($mappingType == 'Search Builder') {
-      $name = "Builder";
       $columnCount = $columnNo;
     }
+    $name = $form->getAttribute('id');
 
     //get the saved mapping details
 
diff -urN civicrm-4.2.19/CRM/Export/BAO/Export.php civicrm-4.2.19-wmf/CRM/Export/BAO/Export.php
--- civicrm-4.2.19/CRM/Export/BAO/Export.php	2014-09-09 15:40:20.000000000 -0700
+++ civicrm-4.2.19-wmf/CRM/Export/BAO/Export.php	2015-03-11 13:31:10.346248475 -0700
@@ -59,6 +59,8 @@
    * @param string $componentTable component table
    * @param bool   $mergeSameAddress merge records if they have same address
    * @param bool   $mergeSameHousehold merge records if they belong to the same household
+   * @param array  $exportParams options passed from the mapper, also used to cache static data
+   * @param string $queryOperator binary SQL operator used to join WHERE clauses
    *
    * @static
    * @access public
@@ -385,9 +387,8 @@
     }
 
     $exportParams['postal_mailing_export']['temp_columns'] = array();
-    if ($exportParams['exportOption'] == 2 &&
-      isset($exportParams['postal_mailing_export']) &&
-      CRM_Utils_Array::value('postal_mailing_export', $exportParams['postal_mailing_export']) == 1
+    if (CRM_Utils_Array::value('exportOption', $exportParams) == 2 &&
+      CRM_Utils_Array::value('postal_mailing_export', $exportParams) == 1
     ) {
       $postalColumns = array('is_deceased', 'do_not_mail', 'street_address', 'supplemental_address_1');
       foreach ($postalColumns as $column) {
@@ -1060,6 +1061,11 @@
       $sql = "DROP TABLE IF EXISTS {$exportTempTable}";
       CRM_Core_DAO::executeQuery($sql);
 
+      if ($componentTable) {
+        $sql = "DROP TABLE IF EXISTS {$componentTable}";
+        CRM_Core_DAO::executeQuery( $sql );
+      }
+
       CRM_Utils_System::civiExit();
     }
     else {
diff -urN civicrm-4.2.19/CRM/Export/Form/Map.php civicrm-4.2.19-wmf/CRM/Export/Form/Map.php
--- civicrm-4.2.19/CRM/Export/Form/Map.php	2014-09-09 15:40:20.000000000 -0700
+++ civicrm-4.2.19-wmf/CRM/Export/Form/Map.php	2015-03-11 13:31:10.346248475 -0700
@@ -152,7 +152,11 @@
    */
   public function postProcess() {
     $params = $this->controller->exportValues($this->_name);
-    $exportParams = $this->controller->exportValues('Select');
+    if (array_key_exists('Select', $this->controller->_pages)) {
+      $exportParams = $this->controller->exportValues('Select');
+    } else {
+      $exportParams = array();
+    }
 
     $greetingOptions = CRM_Export_Form_Select::getGreetingOptions();
 
