diff -urN civicrm-4.2.19/CRM/Contact/Form/Search/Custom/DateAdded.php civicrm-4.2.19-wmf/CRM/Contact/Form/Search/Custom/DateAdded.php
--- civicrm-4.2.19/CRM/Contact/Form/Search/Custom/DateAdded.php	2014-09-09 15:40:20.000000000 -0700
+++ civicrm-4.2.19-wmf/CRM/Contact/Form/Search/Custom/DateAdded.php	2015-03-11 13:31:10.234248478 -0700
@@ -121,7 +121,6 @@
 
     if ($justIDs) {
       $select = "contact_a.id as contact_id";
-    } else {
       $selectClause = "contact_a.id  as contact_id,
                        contact_a.contact_type as contact_type,
                        contact_a.sort_name    as sort_name,
