diff -urN civicrm-4.2.19/CRM/Core/DAO/Address.php civicrm-4.2.19-wmf/CRM/Core/DAO/Address.php

Status: review postal_code requirement
Background: 2 patches in here - one for length of email & one for postal_code. The email one is in core for 4.4 and 4.6.
However the postal code is not - if there are legit postal_codes over 12 char then we just need to get the example and
we can upstream it. Stackflow discussion.
http://stackoverflow.com/questions/325041/i-need-to-store-postal-codes-in-a-database-how-big-should-the-column-be
JIRA: CRM-17118  Bug CRM-13326

--- civicrm-4.2.19/CRM/Core/DAO/Address.php	2014-09-09 15:41:52.000000000 -0700
+++ civicrm-4.2.19-wmf/CRM/Core/DAO/Address.php	2015-03-11 13:31:10.286248477 -0700
@@ -459,12 +459,11 @@
                     'name' => 'postal_code',
                     'type' => CRM_Utils_Type::T_STRING,
                     'title' => ts('Postal Code') ,
-                    'maxlength' => 12,
-                    'size' => CRM_Utils_Type::TWELVE,
+                    'maxlength' => 64,
+                    'size' => CRM_Utils_Type::BIG,
                     'import' => true,
                     'where' => 'civicrm_address.postal_code',
                     'headerPattern' => '/postal|zip/i',
-                    'dataPattern' => '/\d?\d{4}(-\d{4})?/',
                     'export' => true,
                 ) ,
                 'usps_adc' => array(
diff -urN civicrm-4.2.19/CRM/Core/DAO/Email.php civicrm-4.2.19-wmf/CRM/Core/DAO/Email.php
--- civicrm-4.2.19/CRM/Core/DAO/Email.php	2014-09-09 15:41:44.000000000 -0700
+++ civicrm-4.2.19-wmf/CRM/Core/DAO/Email.php	2015-03-11 13:31:10.286248477 -0700
@@ -207,8 +207,8 @@
                     'name' => 'email',
                     'type' => CRM_Utils_Type::T_STRING,
                     'title' => ts('Email') ,
-                    'maxlength' => 64,
-                    'size' => CRM_Utils_Type::BIG,
+                    'maxlength' => 254,
+                    'size' => CRM_Utils_Type::HUGE,
                     'import' => true,
                     'where' => 'civicrm_email.email',
                     'headerPattern' => '/e.?mail/i',
