diff -urN civicrm-4.2.19/CRM/Core/BAO/EntityTag.php civicrm-4.2.19-wmf/CRM/Core/BAO/EntityTag.php
--- civicrm-4.2.19/CRM/Core/BAO/EntityTag.php	2014-09-09 15:40:20.000000000 -0700
+++ civicrm-4.2.19-wmf/CRM/Core/BAO/EntityTag.php	2015-03-11 13:31:10.274248477 -0700
@@ -64,6 +64,42 @@
     return $tags;
   }
 
+    /**
+     *
+     * Given a contact id, it returns an array of tag details the
+     * contact belongs to.
+     *
+     * @param int $entityID id of the entity usually the contactID.
+     * @param string $entityTable name of the entity table usually 'civicrm_contact'
+     *
+     * @return array reference with keys being the tag id
+     *
+     * @access public
+     * @static
+     */
+    static
+    function &getTagDetails($entityID, $entityTable = 'civicrm_contact') {
+      $tags = array();
+
+      $entityTag = new CRM_Core_BAO_EntityTag();
+      $entityTag->entity_id = $entityID;
+      $entityTag->entity_table = $entityTable;
+      $entityTag->joinAdd(new CRM_Core_BAO_Tag());
+      $entityTag->find();
+
+      while ($entityTag->fetch()) {
+        $tags[$entityTag->tag_id] = array(
+          'tag_id' => $entityTag->tag_id,
+          'tag_name' => $entityTag->name,
+          'tag_desc' => $entityTag->description,
+          'tag_parent_id' => $entityTag->parent_id,
+          'entity_table' => $entityTag->entity_table,
+          'entity_id' => $entityTag->entity_id
+        );
+      }
+      return $tags;
+    }
+
   /**
    * takes an associative array and creates a entityTag object
    *
