diff -urN civicrm-4.2.19/CRM/Core/BAO/CustomField.php civicrm-4.2.19-wmf/CRM/Core/BAO/CustomField.php
--- civicrm-4.2.19/CRM/Core/BAO/CustomField.php	2014-09-09 15:40:20.000000000 -0700
+++ civicrm-4.2.19-wmf/CRM/Core/BAO/CustomField.php	2015-03-11 13:31:10.274248477 -0700
@@ -275,7 +275,9 @@
       self::createField($customField, 'modify', $indexExist);
     }
     else {
-      $customField->column_name .= "_{$customField->id}";
+      if (!isset($params['column_name'])) {
+        $customField->column_name .= "_{$customField->id}";
+      }
       $customField->save();
       // make sure all values are present in the object
       $customField->find(TRUE);
@@ -1517,10 +1519,11 @@
       return;
     }
 
+    // WMF: this was a bad feature, has been removed in recent versions of Civi
     // return if field is a 'code' field
-    if (CRM_Utils_Array::value('is_view', $customFields[$customFieldId])) {
-      return;
-    }
+    // if (CRM_Utils_Array::value('is_view', $customFields[$customFieldId])) {
+    //   return;
+    // }
 
     list($tableName, $columnName, $groupID) = self::getTableColumnGroup($customFieldId);
 
@@ -1806,7 +1809,7 @@
       $params['fk_field_name'] = 'id';
       $params['fk_attributes'] = 'ON DELETE SET NULL';
     }
-    if ($field->default_value) {
+    if (isset($field->default_value)) {
       $params['default'] = "'{$field->default_value}'";
     }
 
diff -urN civicrm-4.2.19/CRM/Core/BAO/CustomGroup.php civicrm-4.2.19-wmf/CRM/Core/BAO/CustomGroup.php
--- civicrm-4.2.19/CRM/Core/BAO/CustomGroup.php	2014-09-09 15:40:20.000000000 -0700
+++ civicrm-4.2.19-wmf/CRM/Core/BAO/CustomGroup.php	2015-03-11 13:31:10.274248477 -0700
@@ -142,7 +142,12 @@
 
 
       // lets create the table associated with the group and save it
-      $tableName = $group->table_name = "civicrm_value_" . strtolower(CRM_Utils_String::munge($group->title, '_', 32));
+      if (isset($params['table_name'])) {
+          $tableName = $params['table_name'];
+      } else {
+          $munged_title = strtolower(CRM_Utils_String::munge($group->title, '_', 32));
+          $tableName = "civicrm_value_{$munged_title}_{$group->id}";
+      }
 
       // we do this only once, so name never changes
       $group->name = CRM_Utils_String::munge($params['title'], '_', 64);
@@ -159,9 +164,6 @@
 
     $group->save();
     if ($tableName) {
-      // now append group id to table name, this prevent any name conflicts
-      // like CRM-2742
-      $tableName .= "_{$group->id}";
       $group->table_name = $tableName;
       CRM_Core_DAO::setFieldValue('CRM_Core_DAO_CustomGroup',
         $group->id,
