diff -urN civicrm-4.2.19/CRM/Contribute/Form/Contribution.php civicrm-4.2.19-wmf/CRM/Contribute/Form/Contribution.php
--- civicrm-4.2.19/CRM/Contribute/Form/Contribution.php	2014-09-09 15:40:20.000000000 -0700
+++ civicrm-4.2.19-wmf/CRM/Contribute/Form/Contribution.php	2015-03-11 13:31:10.258248477 -0700
@@ -1044,7 +1044,7 @@
     );
     $this->assign('outBound_option', $mailingInfo['outBound_option']);
 
-    $this->addButtons(array(
+    $buttons = array(
         array(
           'type' => 'upload',
           'name' => ts('Save'),
@@ -1061,9 +1061,17 @@
           'type' => 'cancel',
           'name' => ts('Cancel'),
         ),
-      )
     );
 
+    if ($this->_id) {
+      $buttons[] = array (
+        'type' => 'refund',
+        'js' => array( 'onclick' => "document.location = '" . CRM_Utils_System::url( 'civicrm/contribute/refund', "id={$this->_id}", false, null, false ) . "'; return false;" ),
+        'name' => ts('Refund'),
+      ); 
+    }
+    $this->addButtons($buttons);
+
     $this->addFormRule(array('CRM_Contribute_Form_Contribution', 'formRule'), $this);
 
     if ($this->_action & CRM_Core_Action::VIEW) {
diff -urN civicrm-4.2.19/CRM/Contribute/Form/Refund.php civicrm-4.2.19-wmf/CRM/Contribute/Form/Refund.php
--- civicrm-4.2.19/CRM/Contribute/Form/Refund.php	1969-12-31 16:00:00.000000000 -0800
+++ civicrm-4.2.19-wmf/CRM/Contribute/Form/Refund.php	2015-03-11 13:31:10.262248477 -0700
@@ -0,0 +1,99 @@
+<?php
+
+require_once 'CRM/Contribute/Form.php';
+
+class CRM_Contribute_Form_Refund extends CRM_Contribute_Form
+{
+    public function preProcess() 
+    {
+        parent::preProcess();
+
+        $this->_id = CRM_Utils_Request::retrieve( 'id', 'Positive', $this );
+
+        //$session = CRM_Core_Session::singleton();
+        //$this->_userContext = $session->readUserContext( );
+
+        require_once 'CRM/Contribute/BAO/Contribution.php';
+        $this->contribution = new CRM_Contribute_BAO_Contribution();
+        $this->contribution->id = $this->_id;
+        if ( !$this->contribution->find( true ) ) {
+            CRM_Core_Error::fatal( ts("Contribution does not exist: %1", array( 1 => $this->_id ) ) );
+        }
+
+        require_once 'CRM/Contact/BAO/Contact.php';
+        $this->contact = new CRM_Contact_BAO_Contact();
+        $this->contact->id = $this->contribution->contact_id;
+        $this->contact->find( true );
+    }
+
+    function setDefaultValues( ) {
+        $defaults['type'] = "refund";
+        $defaults['completed'] = true;
+
+        return $defaults;
+    }
+
+
+    /**
+     * Function to build the form
+     *
+     * @return None
+     * @access public
+     */
+    public function buildQuickForm( ) 
+    {
+        $this->addButtons(array(
+            array (
+                'type'      => 'next',
+                'name'      => ts('Confirm')
+            ),
+            array (
+                'type'      => 'cancel',
+                'name'      => ts('Cancel')
+            ),
+        ));
+
+        $this->addElement( 'checkbox', 'completed', ts('Refund has been completed') );
+
+        $this->add('select', 'type', ts('Refund type'),array(
+            '' => '- select -',
+            'refund' => 'Refund',
+            'chargeback' => 'Chargeback',
+        ));
+
+        $this->add( 'hidden', 'id', $this->_id );
+        $this->assign( "contribution_id", $this->_id );
+
+        list( $original_currency, $original_amount ) = explode( " ", $this->contribution->source );
+        $this->assign( "original_currency", $original_currency );
+        $this->assign( "original_amount", $original_amount );
+        $this->assign( "receive_date", $this->contribution->receive_date );
+
+        $this->assign( "contact_name", $this->contact->display_name );
+        $this->assign( "view_original_contribution", CRM_Utils_System::url( 'civicrm/contact/view/contribution', "reset=1&action=view&id={$this->_id}&cid={$this->contact->id}" ) );
+    }
+
+    public function postProcess() 
+    {
+        $submittedValues = $this->controller->exportValues( $this->_name );
+
+        $this->type = CRM_Utils_Array::value( 'type', $submittedValues, 'refund' );
+        $this->completed = CRM_Utils_Array::value( 'completed', $submittedValues, false );
+
+        if ( CRM_Utils_Array::value( '_qf_Refund_next', $this->_submitValues ) ) {
+            try {
+                $this->refund_id = module_invoke( 'wmf_civicrm', 'mark_refund', $this->_id, $this->type, $this->completed );
+            } catch (Exception $e) {
+                CRM_Core_Error::fatal( ts("Failed to mark refund: %1", array( 1 => $e->getMessage() ) ) );
+            }
+
+            CRM_Core_Session::setStatus( ts("Your %1 has been recorded. <a href=\"%2\">Edit</a> the refund contribution.", array(
+                    1 => $this->type,
+                    2 => CRM_Utils_System::url( 'civicrm/contact/view/contribution', "reset=1&action=update&id={$this->refund_id}&cid={$this->contact->id}" ),
+            )) );
+        }
+
+        $session = CRM_Core_Session::singleton();
+        $session->replaceUserContext( CRM_Utils_System::url( 'civicrm/contact/view', "reset=1&cid={$this->contact->id}&selectedChild=contribute" ) );
+    }
+}
diff -urN civicrm-4.2.19/CRM/Contribute/xml/Menu/Contribute.xml civicrm-4.2.19-wmf/CRM/Contribute/xml/Menu/Contribute.xml
--- civicrm-4.2.19/CRM/Contribute/xml/Menu/Contribute.xml	2014-09-09 15:40:20.000000000 -0700
+++ civicrm-4.2.19-wmf/CRM/Contribute/xml/Menu/Contribute.xml	2015-03-11 13:31:10.270248477 -0700
@@ -219,6 +219,11 @@
     <access_arguments>make online contributions</access_arguments>
   </item>
   <item>
+     <path>civicrm/contribute/refund</path>
+     <page_callback>CRM_Contribute_Form_Refund</page_callback>
+     <access_arguments>edit contributions</access_arguments>
+  </item>
+  <item>
     <path>civicrm/contribute/updatebilling</path>
     <title>Update Billing Details</title>
     <page_callback>CRM_Contribute_Form_UpdateBilling</page_callback>
diff -urN civicrm-4.2.19/templates/CRM/Contribute/Form/Refund.tpl civicrm-4.2.19-wmf/templates/CRM/Contribute/Form/Refund.tpl
--- civicrm-4.2.19/templates/CRM/Contribute/Form/Refund.tpl	1969-12-31 16:00:00.000000000 -0800
+++ civicrm-4.2.19-wmf/templates/CRM/Contribute/Form/Refund.tpl	2015-03-11 13:31:10.690248466 -0700
@@ -0,0 +1,31 @@
+{* this template is used for making refund/chargeback contributions *}
+<fieldset><legend>Refund contribution</legend>
+<div class="crm-submit-buttons">{include file="CRM/common/formButtons.tpl" location="top"}</div>
+      <div>
+          <div class="icon inform-icon"></div>
+          {ts}Please complete to record a refund of:{/ts}
+          <div style="border: 1px solid black; margin: 10px; padding: 10px;">
+              {$contact_name}, {$original_currency} {$original_amount} on {$receive_date} (<a href="{$view_original_contribution}">View</a>)
+          </div>
+      </div>
+  <table class="form-layout-compressed">
+     <tr class="crm-contribution-form-block-type">
+	<td class="label">{$form.completed.label}
+	</td>
+	<td class="html-adjust">{$form.completed.html}
+	</td>
+     </tr>
+     <tr class="crm-contribution-form-block-type">
+	<td class="label">{$form.type.label}
+	</td>
+	<td class="html-adjust">{$form.type.html}
+	</td>
+     </tr>
+  </table>
+<div class="crm-submit-buttons">{include file="CRM/common/formButtons.tpl" location="bottom"}</div>
+</fieldset>
+
+<script type="text/javascript">
+{literal}
+{/literal}
+</script>
