diff -urN civicrm-4.2.19/CRM/Contact/Page/DedupeFind.php civicrm-4.2.19-wmf/CRM/Contact/Page/DedupeFind.php

Status: Possible extension w form hook?
Background: This basically blocks deduping except from 2 contacts for performance reasons.

--- civicrm-4.2.19/CRM/Contact/Page/DedupeFind.php	2014-09-09 15:40:20.000000000 -0700
+++ civicrm-4.2.19-wmf/CRM/Contact/Page/DedupeFind.php	2015-03-11 13:31:10.246248478 -0700
@@ -167,6 +167,12 @@
       $where = "de.id IS NULL";
       $this->_mainContacts = CRM_Core_BAO_PrevNextCache::retrieve($cacheKeyString, $join, $where);
       if (empty($this->_mainContacts)) {
+
+if ( empty( $contactIds ) ) {
+    CRM_Core_Session::singleton()->setStatus( "I'm sorry Dave, I can't do that right now.  Crazy deduping of the entire database is not supported at the moment.  Please try merging contacts from the Search Results page, if are feeling that kind of itch." );
+    CRM_Utils_System::redirect( CRM_Utils_System::url( 'civicrm' ) );
+}
+
         if ($gid) {
           $foundDupes = $this->get("dedupe_dupes_$gid");
           if (!$foundDupes) {
