diff -urN civicrm-4.2.19/CRM/Admin/Page/Admin.php civicrm-4.2.19-wmf/CRM/Admin/Page/Admin.php
--- civicrm-4.2.19/CRM/Admin/Page/Admin.php	2014-09-09 15:40:20.000000000 -0700
+++ civicrm-4.2.19-wmf/CRM/Admin/Page/Admin.php	2015-03-11 13:31:10.194248479 -0700
@@ -113,6 +113,9 @@
     $this->assign('localVersion', $versionCheck->localVersion);
     $this->assign('adminPanel', $adminPanel);
 
+    // CRM-14091: Additional sanity checks.
+    CRM_Utils_Check_Security::singleton()->checkAll();
+
     $this->_showHide->addToTemplate();
     return parent::run();
   }
