diff -urN civicrm-4.2.19/CRM/Contact/Form/CustomData.php civicrm-4.2.19-wmf/CRM/Contact/Form/CustomData.php
--- civicrm-4.2.19/CRM/Contact/Form/CustomData.php	2014-09-09 15:40:20.000000000 -0700
+++ civicrm-4.2.19-wmf/CRM/Contact/Form/CustomData.php	2015-03-11 13:31:10.230248478 -0700
@@ -122,7 +122,7 @@
       return CRM_Custom_Form_CustomData::preProcess($this);
     }
 
-    $this->_groupID = CRM_Utils_Request::retrieve('groupID', 'Positive', $this, TRUE);
+    $this->_groupID = CRM_Utils_Request::retrieve('groupId', 'Positive', $this, TRUE);
     $this->_tableID = CRM_Utils_Request::retrieve('tableId', 'Positive', $this, TRUE);
 
     $this->_contactType = CRM_Contact_BAO_Contact::getContactType($this->_tableID);
diff -urN civicrm-4.2.19/templates/CRM/Custom/Page/CustomDataView.tpl civicrm-4.2.19-wmf/templates/CRM/Custom/Page/CustomDataView.tpl
--- civicrm-4.2.19/templates/CRM/Custom/Page/CustomDataView.tpl	2014-09-09 15:40:20.000000000 -0700
+++ civicrm-4.2.19-wmf/templates/CRM/Custom/Page/CustomDataView.tpl	2015-03-11 13:31:10.694248466 -0700
@@ -32,7 +32,7 @@
       {if $editOwnCustomData or ($showEdit and $editCustomData and $groupId)}
         <tr>
           <td>
-            <a href="{crmURL p="civicrm/contact/view/cd/edit" q="tableId=`$contactId`&cid=`$contactId`&groupID=`$groupId`&action=update&reset=1"}" class="button" style="margin-left: 6px;"><span><div class="icon edit-icon"></div>{ts 1=$cd_edit.title}Edit %1{/ts}</span></a><br/><br/>
+            <a href="{crmURL p="civicrm/contact/view/cd/edit" q="tableId=`$contactId`&cid=`$contactId`&groupId=`$groupId`&action=update&reset=1"}" class="button" style="margin-left: 6px;"><span><div class="icon edit-icon"></div>{ts 1=$cd_edit.title}Edit %1{/ts}</span></a><br/><br/>
           </td>
         </tr>
       {/if}
