diff -urN civicrm-4.2.19/CRM/ACL/BAO/Cache.php civicrm-4.2.19-wmf/CRM/ACL/BAO/Cache.php

Status: Needs investigation
Background: Patch changes various places from TRUNCATE to DELETE FROM
JIRA: CRM-10700

--- civicrm-4.2.19/CRM/ACL/BAO/Cache.php	2014-09-09 15:40:20.000000000 -0700
+++ civicrm-4.2.19-wmf/CRM/ACL/BAO/Cache.php	2015-03-11 13:31:10.174248480 -0700
@@ -149,7 +149,7 @@
 ";
     CRM_Core_DAO::singleValueQuery($query);
 
-    CRM_Core_DAO::singleValueQuery("TRUNCATE TABLE civicrm_acl_contact_cache");
+    CRM_Core_DAO::singleValueQuery("DELETE FROM civicrm_acl_contact_cache");
   }
 }
 
diff -urN civicrm-4.2.19/CRM/Contact/BAO/GroupContactCache.php civicrm-4.2.19-wmf/CRM/Contact/BAO/GroupContactCache.php
--- civicrm-4.2.19/CRM/Contact/BAO/GroupContactCache.php	2014-09-09 15:40:20.000000000 -0700
+++ civicrm-4.2.19-wmf/CRM/Contact/BAO/GroupContactCache.php	2015-03-11 13:31:10.222248478 -0700
@@ -195,7 +195,7 @@
 
       if ($smartGroupCacheTimeout == 0) {
         $query = "
-TRUNCATE civicrm_group_contact_cache
+DELETE FROM civicrm_group_contact_cache
 ";
         $update = "
 UPDATE civicrm_group g
diff -urN civicrm-4.2.19/CRM/Contact/Form/Search/Custom/FullText.php civicrm-4.2.19-wmf/CRM/Contact/Form/Search/Custom/FullText.php
--- civicrm-4.2.19/CRM/Contact/Form/Search/Custom/FullText.php	2014-09-09 15:40:20.000000000 -0700
+++ civicrm-4.2.19-wmf/CRM/Contact/Form/Search/Custom/FullText.php	2015-03-11 13:31:10.238248478 -0700
@@ -345,7 +345,7 @@
   }
 
   function runQueries(&$tables) {
-    $sql = "TRUNCATE {$this->_entityIDTableName}";
+    $sql = "DELETE FROM {$this->_entityIDTableName}";
     CRM_Core_DAO::executeQuery($sql);
 
     $maxRowCount = 0;
diff -urN civicrm-4.2.19/CRM/Core/Config.php civicrm-4.2.19-wmf/CRM/Core/Config.php
--- civicrm-4.2.19/CRM/Core/Config.php	2014-09-09 15:40:20.000000000 -0700
+++ civicrm-4.2.19-wmf/CRM/Core/Config.php	2015-03-11 13:31:10.286248477 -0700
@@ -689,13 +689,13 @@
    */
   function clearDBCache() {
     $queries = array(
-      'TRUNCATE TABLE civicrm_acl_cache',
-      'TRUNCATE TABLE civicrm_acl_contact_cache',
-      'TRUNCATE TABLE civicrm_cache',
-      'TRUNCATE TABLE civicrm_prevnext_cache',
+      'DELETE FROM civicrm_acl_cache',
+      'DELETE FROM civicrm_acl_contact_cache',
+      'DELETE FROM civicrm_cache',
+      'DELETE FROM civicrm_prevnext_cache',
       'UPDATE civicrm_group SET cache_date = NULL',
-      'TRUNCATE TABLE civicrm_group_contact_cache',
-      'TRUNCATE TABLE civicrm_menu',
+      'DELETE FROM civicrm_group_contact_cache',
+      'DELETE FROM civicrm_menu',
       'UPDATE civicrm_setting SET value = NULL WHERE name="navigation" AND contact_id IS NOT NULL',
       'DELETE FROM civicrm_setting WHERE name="modulePaths"', // CRM-10543
     );
diff -urN civicrm-4.2.19/CRM/Core/Menu.php civicrm-4.2.19-wmf/CRM/Core/Menu.php
--- civicrm-4.2.19/CRM/Core/Menu.php	2014-09-09 15:40:20.000000000 -0700
+++ civicrm-4.2.19-wmf/CRM/Core/Menu.php	2015-03-11 13:31:10.298248476 -0700
@@ -235,7 +235,7 @@
   static function store($truncate = TRUE) {
     // first clean up the db
     if ($truncate) {
-      $query = 'TRUNCATE civicrm_menu';
+      $query = 'DELETE FROM civicrm_menu';
       CRM_Core_DAO::executeQuery($query);
     }
     $menuArray = self::items();
