diff -urN civicrm-4.2.19/CRM/Core/DAO.php civicrm-4.2.19-wmf/CRM/Core/DAO.php
--- civicrm-4.2.19/CRM/Core/DAO.php	2014-09-09 15:40:20.000000000 -0700
+++ civicrm-4.2.19-wmf/CRM/Core/DAO.php	2015-03-11 13:31:10.286248477 -0700
@@ -1475,7 +1475,7 @@
     }
 
     if ($trigger) {
-      $dao->query('CREATE TRIGGER civicrm_domain_trigger BEFORE INSERT ON civicrm_domain FOR EACH ROW BEGIN END');
+      //$dao->query('CREATE TRIGGER civicrm_domain_trigger BEFORE INSERT ON civicrm_domain FOR EACH ROW BEGIN END');
 
       if (PEAR::getStaticProperty('DB_DataObject', 'lastError')) {
         CRM_Core_Error::setCallback();
@@ -1485,7 +1485,7 @@
         return FALSE;
       }
 
-      $dao->query('DROP TRIGGER IF EXISTS civicrm_domain_trigger');
+      //$dao->query('DROP TRIGGER IF EXISTS civicrm_domain_trigger');
       if (PEAR::getStaticProperty('DB_DataObject', 'lastError')) {
         CRM_Core_Error::setCallback();
         if ($view) {
@@ -1625,6 +1625,7 @@
           $triggerName = "{$tableName}_{$whenName}_{$eventName}";
           $triggerSQL  = "CREATE TRIGGER $triggerName $whenName $eventName ON $tableName FOR EACH ROW BEGIN $varString $sqlString END";
 
+		  /*
           CRM_Core_DAO::executeQuery("DROP TRIGGER IF EXISTS $triggerName");
           CRM_Core_DAO::executeQuery($triggerSQL,
             array(),
@@ -1633,6 +1634,7 @@
             FALSE,
             FALSE
           );
+		  */
         }
       }
     }
diff -urN civicrm-4.2.19/CRM/Core/I18n/Schema.php civicrm-4.2.19-wmf/CRM/Core/I18n/Schema.php
--- civicrm-4.2.19/CRM/Core/I18n/Schema.php	2014-09-09 15:40:20.000000000 -0700
+++ civicrm-4.2.19-wmf/CRM/Core/I18n/Schema.php	2015-03-11 13:31:10.294248476 -0700
@@ -220,7 +220,7 @@
           $when = $triggerInfo['when'];
           $event = $triggerInfo['event'];
           $triggerName = "{$table}_{$when}_{$event}";
-          CRM_Core_DAO::executeQuery("DROP TRIGGER IF EXISTS {$triggerName}");
+          //CRM_Core_DAO::executeQuery("DROP TRIGGER IF EXISTS {$triggerName}");
         }
       }
 
diff -urN civicrm-4.2.19/CRM/Logging/Schema.php civicrm-4.2.19-wmf/CRM/Logging/Schema.php
--- civicrm-4.2.19/CRM/Logging/Schema.php	2014-09-09 15:40:20.000000000 -0700
+++ civicrm-4.2.19-wmf/CRM/Logging/Schema.php	2015-03-11 13:31:10.358248475 -0700
@@ -120,14 +120,14 @@
 
     foreach ($tableNames as $table) {
       // before triggers
-      $dao->executeQuery("DROP TRIGGER IF EXISTS {$table}_before_insert");
-      $dao->executeQuery("DROP TRIGGER IF EXISTS {$table}_before_update");
-      $dao->executeQuery("DROP TRIGGER IF EXISTS {$table}_before_delete");
+      //error_log("DROP TRIGGER IF EXISTS {$table}_before_insert");
+      //error_log("DROP TRIGGER IF EXISTS {$table}_before_update");
+      //error_log("DROP TRIGGER IF EXISTS {$table}_before_delete");
 
      // after triggers
-      $dao->executeQuery("DROP TRIGGER IF EXISTS {$table}_after_insert");
-      $dao->executeQuery("DROP TRIGGER IF EXISTS {$table}_after_update");
-      $dao->executeQuery("DROP TRIGGER IF EXISTS {$table}_after_delete");
+      //error_log("DROP TRIGGER IF EXISTS {$table}_after_insert");
+      //error_log("DROP TRIGGER IF EXISTS {$table}_after_update");
+      //error_log("DROP TRIGGER IF EXISTS {$table}_after_delete");
     }
   }
 
