diff -urN civicrm-4.2.19/CRM/Contribute/Form/Search.php civicrm-4.2.19-wmf/CRM/Contribute/Form/Search.php
--- civicrm-4.2.19/CRM/Contribute/Form/Search.php	2014-09-09 15:40:20.000000000 -0700
+++ civicrm-4.2.19-wmf/CRM/Contribute/Form/Search.php	2015-03-11 13:31:10.262248477 -0700
@@ -159,6 +159,13 @@
     $this->_limit   = CRM_Utils_Request::retrieve('limit', 'Positive', $this);
     $this->_context = CRM_Utils_Request::retrieve('context', 'String', $this, FALSE, 'search');
 
+    /*
+     * WMF HACK: "force" causes a search with null criteria.  Disable until this is fixed.
+     */
+    if ( $this->_context === "search" && CRM_Utils_Request::retrieve('qfKey', 'String') === NULL ) {
+      $this->_force = false;
+    }
+
     $this->assign("context", $this->_context);
 
     // get user submitted values
