diff -urN civicrm-4.2.19/CRM/Contribute/BAO/ContributionRecur.php civicrm-4.2.19-wmf/CRM/Contribute/BAO/ContributionRecur.php

Status: needs upstreaming
Background: patch appears to add failure_count & failure_retry_date to recurring contribution tab.

--- civicrm-4.2.19/CRM/Contribute/BAO/ContributionRecur.php	2014-09-09 15:40:20.000000000 -0700
+++ civicrm-4.2.19-wmf/CRM/Contribute/BAO/ContributionRecur.php	2015-03-11 13:31:10.254248478 -0700
@@ -338,6 +338,8 @@
       $params[$recurDAO->id]['next_sched_contribution'] = $recurDAO->next_sched_contribution;
       $params[$recurDAO->id]['amount'] = $recurDAO->amount;
       $params[$recurDAO->id]['currency'] = $recurDAO->currency;
+      $params[$recurDAO->id]['failure_count'] = $recurDAO->failure_count;
+      $params[$recurDAO->id]['failure_retry_date'] = $recurDAO->failure_retry_date;
       $params[$recurDAO->id]['frequency_unit'] = $recurDAO->frequency_unit;
       $params[$recurDAO->id]['frequency_interval'] = $recurDAO->frequency_interval;
       $params[$recurDAO->id]['installments'] = $recurDAO->installments;
