diff -urN civicrm-4.2.19/CRM/Contact/BAO/Contact.php civicrm-4.2.19-wmf/CRM/Contact/BAO/Contact.php

Status: needs more information.
Background: patch causes this line to be skipped on create  NOT: self::processGreetings($contact);
The patch has been through a few iterations.... No explanation.

--- civicrm-4.2.19/CRM/Contact/BAO/Contact.php	2014-09-09 15:40:20.000000000 -0700
+++ civicrm-4.2.19-wmf/CRM/Contact/BAO/Contact.php	2015-03-11 13:31:10.218248478 -0700
@@ -426,7 +428,7 @@
     }
 
     // process greetings CRM-4575, cache greetings
-    self::processGreetings($contact);
+    // NOT: self::processGreetings($contact);
 
     return $contact;
   }
